src/app/fdctl/run/tiles/fd_repair.c

```diff
@@ -0,0 +1,675 @@
+  }
+
+  fd_stake_weight_t const * in_stake_weights = fd_type_pun_const( header+4UL );
+  fd_repair_set_stake_weights( ctx->repair, in_stake_weights, stakes_cnt );
+}
+
+
+static void 
+repair_send_packet( uchar const * msg, 
+                    size_t msglen, 
+                    fd_gossip_peer_addr_t const * addr, 
+                    void * arg ) {
+  ulong tsorig = fd_frag_meta_ts_comp( fd_tickcount() );
+  send_packet( arg, addr->addr, fd_ushort_bswap( addr->port ), msg, msglen, tsorig );
+}
+
+static void
+repair_shred_deliver( fd_shred_t const * shred,
+                      ulong shred_sz, 
+                      fd_repair_peer_addr_t const * from FD_PARAM_UNUSED, 
+                      fd_pubkey_t const * id FD_PARAM_UNUSED, 
+                      void * arg ) {
+  fd_repair_tile_ctx_t * ctx = (fd_repair_tile_ctx_t *)arg;
+  ulong tsorig = fd_frag_meta_ts_comp( fd_tickcount() );
+  
+  fd_shred_t * out_shred = fd_chunk_to_laddr( ctx->store_out_mem, ctx->store_out_chunk );
+  fd_memcpy( out_shred, shred, shred_sz );
+  
+  ulong tspub = fd_frag_meta_ts_comp( fd_tickcount() );
+  fd_mcache_publish( ctx->store_out_mcache, ctx->store_out_depth, ctx->store_out_seq, 1, ctx->store_out_chunk,
+    shred_sz, 0UL, tsorig, tspub );
+  ctx->store_out_seq   = fd_seq_inc( ctx->store_out_seq, 1UL );
+  ctx->store_out_chunk = fd_dcache_compact_next( ctx->store_out_chunk, shred_sz, ctx->store_out_chunk0, ctx->store_out_wmark );}
+
+static void
+repair_shred_deliver_fail( fd_pubkey_t const * id FD_PARAM_UNUSED, 
+                           ulong slot, 
+                           uint shred_index,
+                           void * arg FD_PARAM_UNUSED,
+                           int reason ) {
+  FD_LOG_WARNING(( "repair failed to get shred - slot: %lu, shred_index: %u, reason: %u", slot, shred_index, reason ));
+}
+
+static void
+before_frag( void * _ctx FD_PARAM_UNUSED,
+             ulong  in_idx,
+             ulong  seq  FD_PARAM_UNUSED,
+             ulong  sig,
+             int *  opt_filter ) {
+
+  if( FD_LIKELY( in_idx==NET_IN_IDX ) ) {
+    *opt_filter = fd_disco_netmux_sig_proto( sig )!=DST_PROTO_REPAIR;
+  }
+}
+
+static void
+during_frag( void * _ctx,
+             ulong  in_idx,
+             ulong  seq FD_PARAM_UNUSED,
+             ulong  sig,
+             ulong  chunk,
+             ulong  sz,
+             int *  opt_filter ) {
+
+  fd_repair_tile_ctx_t * ctx = (fd_repair_tile_ctx_t *)_ctx;
+
+  if( FD_UNLIKELY( in_idx==CONTACT_IN_IDX ) ) {
+    if( FD_UNLIKELY( chunk<ctx->contact_in_chunk0 || chunk>ctx->contact_in_wmark ) )
+      FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz,
+            ctx->contact_in_chunk0, ctx->contact_in_wmark ));
+
+    uchar const * dcache_entry = fd_chunk_to_laddr_const( ctx->contact_in_mem, chunk );
+    handle_new_cluster_contact_info( ctx, dcache_entry );
```

Comment:
We typically don't handle data like this in during_frag, because you can get overrun your consumer does not know how to detect corruption. You want to copy the data out to a buffer, then process it in after_frag. Same for the below cases.